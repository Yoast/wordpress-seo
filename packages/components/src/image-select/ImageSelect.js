import React from "react";
import { __ } from "@wordpress/i18n";
import ImageSelectButtons from "./ImageSelectButtons";
import PropTypes from "prop-types";
import FieldGroup from "../field-group/FieldGroup";
import Alert from "../Alert";

/**
 * Renders ImageSelect component.
 *
 * @param {Object} props The props.
 *
 * @returns {React.Component} The ImageSelect.
 */
function ImageSelect( props ) {
	const imageSelected = props.usingFallback === false && props.imageUrl !== "";
	const previewImageUrl = props.imageUrl || props.defaultImageUrl || "";
	const showWarnings = props.warnings.length > 0 && imageSelected;

	let imageClassName = showWarnings ? "yoast-image-select__preview yoast-image-select__preview-has-warnings" : "yoast-image-select__preview";
	if ( previewImageUrl === "" ) {
		imageClassName = "yoast-image-select__preview yoast-image-select__preview--no-preview";
	}

	const imageSelectButtonsProps = {
		imageSelected: imageSelected,
		onClick: props.onClick,
		onRemoveImageClick: props.onRemoveImageClick,
		selectImageButtonId: props.selectImageButtonId,
		replaceImageButtonId: props.replaceImageButtonId,
		removeImageButtonId: props.removeImageButtonId,
		isDisabled: props.isDisabled,
	};

	/**
	 * @returns {JSXElement} returns a text for screen readers.
	 */
	const ScreenReaderText = () => {
		return (
			<span className="screen-reader-text">
				{
					imageSelected
						? __( "Replace image", "wordpress-seo" )
						: __( "Select image", "wordpress-seo" )
				}
			</span>
		);
	};

	return (
		<div
			className="yoast-image-select"
			onMouseEnter={ props.onMouseEnter }
			onMouseLeave={ props.onMouseLeave }
		>
			<FieldGroup
				label={ props.label }
				hasNewBadge={ props.hasNewBadge }
				hasPremiumBadge={ props.hasPremiumBadge }
			>
				{ props.hasPreview &&
					<button
						className={ imageClassName }
						onClick={ props.onClick }
						type="button"
						disabled={ props.isDisabled }
					>
						{ previewImageUrl !== "" &&
							<img src={ previewImageUrl } alt={ props.imageAltText } className="yoast-image-select__preview--image" />
						}
						<ScreenReaderText />
					</button>
				}
				{
					showWarnings && <div role="alert">
						{
							props.warnings.map( ( warning, index ) => <Alert key={ `warning${ index }` } type="warning">
								{ warning }
							</Alert> )
						}
					</div>
				}
				<ImageSelectButtons { ...imageSelectButtonsProps } />
			</FieldGroup>
		</div>
	);
}

export default ImageSelect;

ImageSelect.propTypes = {
	defaultImageUrl: PropTypes.string,
	imageUrl: PropTypes.string,
	imageAltText: PropTypes.string,
	hasPreview: PropTypes.bool.isRequired,
	label: PropTypes.string.isRequired,
	onClick: PropTypes.func,
	onMouseEnter: PropTypes.func,
	onMouseLeave: PropTypes.func,
	onRemoveImageClick: PropTypes.func,
	selectImageButtonId: PropTypes.string,
	replaceImageButtonId: PropTypes.string,
	removeImageButtonId: PropTypes.string,
	warnings: PropTypes.arrayOf( PropTypes.string ),
	hasNewBadge: PropTypes.bool,
	isDisabled: PropTypes.bool,
	usingFallback: PropTypes.bool,
	hasPremiumBadge: PropTypes.bool,
};

ImageSelect.defaultProps = {
	defaultImageUrl: "",
	imageUrl: "",
	imageAltText: "",
	onClick: () => {},
	onMouseEnter: () => {},
	onMouseLeave: () => {},
	onRemoveImageClick: () => {},
	selectImageButtonId: "",
	replaceImageButtonId: "",
	removeImageButtonId: "",
	warnings: [],
	hasNewBadge: false,
	isDisabled: false,
	usingFallback: false,
	hasPremiumBadge: false,
};
