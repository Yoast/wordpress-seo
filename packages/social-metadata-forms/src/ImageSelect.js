import { __ } from "@wordpress/i18n";
import { Alert, InputField, SimulatedLabel } from "@yoast/components";
import { StandardButton } from "@yoast/replacement-variable-editor";
import PropTypes from "prop-types";
import React, { Fragment } from "react";
import styled from "styled-components";

const UndoButton = styled( StandardButton )`
	background: none!important;
	border: none;
	padding: 0!important;
	color: #a00;
	text-decoration: underline;
	cursor: pointer;
	box-shadow: none;

	&:hover {
		color: #F00;
	}

	&:focus {
		color: #F00;
	}
`;

const UrlInputField = styled( InputField )`
	min-width: 100%;
`;

const ColumnWrapper = styled.div`
	display: flex;
	flex-direction: column;
`;

const RowWrapper = styled.div`
	display: flex;
	margin: 10px 0 0 0;
`;

const DivWithMargin = styled.div`
	margin-top: 24px;
`;

/**
 * Renders the select/replace and remove buttons.
 *
 * @param {object}   props                      The properties passed to this component.
 * @param {bool}     props.imageSelected        Is there already an image selected.
 * @param {function} props.onClick              Callback called when the "Select image" or "Replace image" button is clicked.
 * @param {function} props.onRemoveImageClick   Callback called when the "Remove image" button is clicked.
 * @param {string}   props.selectImageButtonId  The ID for the select image button.
 * @param {string}   props.replaceImageButtonId The ID for the replace image button.
 * @param {string}   props.removeImageButtonId  The ID for the remove image button.
 *
 * @returns {JSX.Element} The buttons to render for the ImageSelect.
 */
const ImageSelectButtons = ( {
	imageSelected,
	onClick,
	onRemoveImageClick,
	selectImageButtonId,
	replaceImageButtonId,
	removeImageButtonId,
} ) => {
	return <Fragment>
		<StandardButton
			id={ imageSelected ? replaceImageButtonId : selectImageButtonId }
			onClick={ onClick }
		>
			{
				imageSelected
					? __( "Replace image", "wordpress-seo" )
					: __( "Select image", "wordpress-seo" )
			}
		</StandardButton>
		{
			imageSelected && <UndoButton
				id={ removeImageButtonId }
				onClick={ onRemoveImageClick }
			>
				{ __( "Remove image", "wordpress-seo" ) }
			</UndoButton>
		}
	</Fragment>;
};

ImageSelectButtons.propTypes = {
	imageSelected: PropTypes.bool.isRequired,
	onClick: PropTypes.func.isRequired,
	onRemoveImageClick: PropTypes.func.isRequired,
	selectImageButtonId: PropTypes.string.isRequired,
	replaceImageButtonId: PropTypes.string.isRequired,
	removeImageButtonId: PropTypes.string.isRequired,
};

/**
 * Component for displaying an image selection button with a title.
 *
 * Displays an warning message when the selected image cannot be used.
 *
 * @param {object}   props                      The properties passed to this component.
 * @param {string}   props.title                The title that is displayed above the selection button.
 * @param {string[]} props.warnings             An array of warnings that detail why the image cannot be used.
 * @param {function} props.onClick              Callback called when the "Select image" or "Replace image" button is clicked.
 * @param {function} props.onRemoveImageClick   Callback called when the "Remove image" button is clicked.
 * @param {string}   props.imageUrl             The Url adress of the image
 * @param {bool}     props.isPremium            States if premium is installed.
 * @param {string}   props.imageUrlInputId      The ID for the image URL input.
 * @param {string}   props.selectImageButtonId  The ID for the select image button.
 * @param {string}   props.replaceImageButtonId The ID for the replace image button.
 * @param {string}   props.removeImageButtonId  The ID for the remove image button.
 *
 * @returns {JSX.Element} The ImageSelect component with a title, optional warnings and an image selection button.
 */
const ImageSelect = ( {
	title,
	warnings,
	imageSelected,
	onClick,
	onRemoveImageClick,
	imageUrl,
	isPremium,
	onMouseEnter,
	onMouseLeave,
	imageUrlInputId,
	selectImageButtonId,
	replaceImageButtonId,
	removeImageButtonId,
} ) => {
	const imageSelectButtonsProps = {
		imageSelected,
		onClick,
		onRemoveImageClick,
		selectImageButtonId,
		replaceImageButtonId,
		removeImageButtonId,
	};

	return <DivWithMargin
		onMouseEnter={ onMouseEnter }
		onMouseLeave={ onMouseLeave }
	>
		<SimulatedLabel>
			{ title }
		</SimulatedLabel>
		{
			warnings.length > 0 && imageSelected &&
			warnings.map( ( warning, index ) => <Alert key={ `warning${ index }` } type="warning">
				{ warning }
			</Alert> )
		}
		{
			isPremium ? <ImageSelectButtons { ...imageSelectButtonsProps } />
				: <ColumnWrapper>
					<UrlInputField
						id={ imageUrlInputId }
						value={ imageUrl }
						disabled={ "disabled" }
					/>
					<RowWrapper>
						<ImageSelectButtons { ...imageSelectButtonsProps } />
					</RowWrapper>
				</ColumnWrapper>
		}
	</DivWithMargin>;
};

ImageSelect.propTypes = {
	title: PropTypes.string.isRequired,
	imageSelected: PropTypes.bool.isRequired,
	isPremium: PropTypes.bool.isRequired,
	warnings: PropTypes.arrayOf( PropTypes.string ),
	imageUrl: PropTypes.string,
	onMouseEnter: PropTypes.func,
	onMouseLeave: PropTypes.func,
	imageUrlInputId: PropTypes.string,
	onClick: PropTypes.func,
	onRemoveImageClick: PropTypes.func,
	selectImageButtonId: PropTypes.string,
	replaceImageButtonId: PropTypes.string,
	removeImageButtonId: PropTypes.string,
};

ImageSelect.defaultProps = {
	warnings: [],
	imageUrl: "",
	onMouseEnter: () => {},
	onMouseLeave: () => {},
	onClick: () => {},
	onRemoveImageClick: () => {},
	imageUrlInputId: "",
	selectImageButtonId: "",
	replaceImageButtonId: "",
	removeImageButtonId: "",
};

export default ImageSelect;
