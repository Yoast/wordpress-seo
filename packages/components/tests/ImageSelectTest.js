import React from "react";
import renderer from "react-test-renderer";

import ImageSelect from "../src/image-select/ImageSelect";

describe( "ImageSelect", () => {
	test( "ImageSelects match the snapshot", () => {
		const imageSelects = [
			{
				label: "Organization",
				hasPreview: true,
				imageUrl: "",
			},
			{
				label: "Another organization",
				hasPreview: true,
				imageUrl: "http://placekitten.com/300/300",
			},
			{
				label: "Yet another organization",
				hasPreview: false,
				imageUrl: "",
			},
			{
				label: "Not another organization",
				hasPreview: false,
				imageUrl: "http://placekitten.com/300/300",
			},
		];

		imageSelects.forEach( imageSelect => {
			const component = renderer.create(
				<ImageSelect
					label={ imageSelect.label }
					imageAltText={ "" }
					hasPreview={ imageSelect.hasPreview }
					imageUrl={ imageSelect.imageUrl }
				/>
			);
			const tree = component.toJSON();

			expect( tree ).toMatchSnapshot();
		} );
	} );
} );
