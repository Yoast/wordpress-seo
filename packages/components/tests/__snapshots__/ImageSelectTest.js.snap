// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSelect ImageSelects match the snapshot 1`] = `
<div
  className="yoast-image-select"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="yoast-field-group"
  >
    <div
      className="yoast-field-group__title"
    >
      <b>
        Organization
      </b>
    </div>
    <button
      className="yoast-image-select__preview yoast-image-select__preview--no-preview"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <span
        className="screen-reader-text"
      >
        Select image
      </span>
    </button>
    <div
      className="yoast-image-select-buttons"
    >
      <button
        className="yoast-button yoast-button--secondary"
        disabled={false}
        id=""
        onClick={[Function]}
        type="button"
      >
        Select image
      </button>
    </div>
  </div>
</div>
`;

exports[`ImageSelect ImageSelects match the snapshot 2`] = `
<div
  className="yoast-image-select"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="yoast-field-group"
  >
    <div
      className="yoast-field-group__title"
    >
      <b>
        Another organization
      </b>
    </div>
    <button
      className="yoast-image-select__preview"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        className="yoast-image-select__preview--image"
        src="http://placekitten.com/300/300"
      />
      <span
        className="screen-reader-text"
      >
        Replace image
      </span>
    </button>
    <div
      className="yoast-image-select-buttons"
    >
      <button
        className="yoast-button yoast-button--secondary"
        disabled={false}
        id=""
        onClick={[Function]}
        type="button"
      >
        Replace image
      </button>
      <button
        className="yoast-remove"
        disabled={false}
        id=""
        onClick={[Function]}
        type="button"
      >
        Remove image
      </button>
    </div>
  </div>
</div>
`;

exports[`ImageSelect ImageSelects match the snapshot 3`] = `
<div
  className="yoast-image-select"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="yoast-field-group"
  >
    <div
      className="yoast-field-group__title"
    >
      <b>
        Yet another organization
      </b>
    </div>
    <div
      className="yoast-image-select-buttons"
    >
      <button
        className="yoast-button yoast-button--secondary"
        disabled={false}
        id=""
        onClick={[Function]}
        type="button"
      >
        Select image
      </button>
    </div>
  </div>
</div>
`;

exports[`ImageSelect ImageSelects match the snapshot 4`] = `
<div
  className="yoast-image-select"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="yoast-field-group"
  >
    <div
      className="yoast-field-group__title"
    >
      <b>
        Not another organization
      </b>
    </div>
    <div
      className="yoast-image-select-buttons"
    >
      <button
        className="yoast-button yoast-button--secondary"
        disabled={false}
        id=""
        onClick={[Function]}
        type="button"
      >
        Replace image
      </button>
      <button
        className="yoast-remove"
        disabled={false}
        id=""
        onClick={[Function]}
        type="button"
      >
        Remove image
      </button>
    </div>
  </div>
</div>
`;
